src/common/protobuf_utils.hpp

```diff
@@ -64,6 +64,63 @@ struct Slave;
 
 namespace protobuf {
 
+//
+// A "protobuf union" is a protobuf message
+// which has an enum named "Type" defined in its scope,
+// a required field named "type" of a "Type" type
+// and a field corresponding to each member of the "Type" enum.
+//
+// A valid "protobuf union" should satisfy the following requirements.
+// 1. A member of the Type enum with a number (not index!) of 0,
+//    if there is any, should be named "UNKNOWN".
+// 2. For each other member of the Type enum there should be an optional field
+//    in the message with an exactly matching name in lowercase.
+// 3. The fields that correspond to members of the Type enum other than
+//    message.type(), should not be set.
+//
+// NOTE: If possible, oneof should be used in the new messages instead
+// of the "protobuf union".
+//
+class UnionValidatorHelper
+{
+public:
+  UnionValidatorHelper(
+      const google::protobuf::Descriptor*,
+      const google::protobuf::EnumDescriptor*);
+
+  Option<Error> Validate(
+      const int messageTypeNumber, const google::protobuf::Message&) const;
+
+private:
+  std::vector<std::pair<int, const google::protobuf::FieldDescriptor*>>
+    UnionFieldDescriptors;
+  const google::protobuf::EnumDescriptor* TypeDescriptor;
+};
+
+
+template <class Message>
+class UnionValidator
+{
+public:
+  static Option<Error> Validate(const Message& message)
+  {
+    static const UnionValidatorHelper helper(
```

Comment:
Couldn't you just move this static variable into `validateProtobufUnion()` and get rid of this class? If yes, we could also rename `UnionValidatorHelper` -> `UnionValidator`.

Reply:
Exactly. We do not need this class at all. Moved and renamed, as you suggested.